### PR TITLE
code move so user file patterns can be resolved in lib crate

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

**Review by martinvonz — COMMENTED** [review]
> It seems a bit invasive that RepoPath constructor processes an environment
like cwd

What do you think about leaving the `file_util::normalize_path(&cwd.join(input))` step outside of `RepoPath`? That seems more consistent with `RepoPath::to_fs_path()`.

**Comment by yuja** [comment]
> What do you think about leaving the `file_util::normalize_path(&cwd.join(input))` step outside of `RepoPath`? That seems more consistent with `RepoPath::to_fs_path()`.

Yeah, that was close to my first attempt, but we would need to `.map_err(|_| FilePathParseError(input.to_owned()))` somewhere if `RepoPath` doesn't know the original input. Another (minor) problem is that if `input` were `&Path`, it could theoretically contain invalid UTF-8 sequence and so the error type would have to be InvalidUtf8|NotInRepo enum, and the `map_err()` would become more verbose. That's why `(cwd, base)` pair is passed into the lib function.

**Comment by martinvonz** [comment]
> Yeah, that was close to my first attempt, but we would need to `.map_err(|_| FilePathParseError(input.to_owned()))` somewhere if `RepoPath` doesn't know the original input. Another (minor) problem is that if `input` were `&Path`, it could theoretically contain invalid UTF-8 sequence and so the error type would have to be InvalidUtf8|NotInRepo enum, and the `map_err()` would become more verbose. That's why `(cwd, base)` pair is passed into the lib function.

That doesn't seem like much of a problem, especially since it seems to be very few places that would be affected. I'll review the rest now and we can always move that bit out in a follow-up if we decide that it's better.

File: lib/src/repo_path.rs [hunk 1]
```diff
@@ -76,6 +78,28 @@ impl RepoPath {
         RepoPath { components }
     }
 
+    /// Parses an `input` path into a `RepoPath` relative to `base`.
+    ///
+    /// The `cwd` and `base` paths are supposed to be absolute and normalized in
+    /// the same manner. The `input` path may be either relative to `cwd` or
+    /// absolute.
+    pub fn parse_fs_path(cwd: &Path, base: &Path, input: &str) -> Result<Self, FsPathParseError> {
+        use std::path::Component;
```
Comment (martinvonz) [inline lib/src/repo_path.rs:87]: Can this be at the top of the file? It doesn't look like it would conflict with anything.

**Review by martinvonz — APPROVED** [review]

**Comment by yuja** [comment]
> especially since it seems to be very few places that would be affected. I'll review the rest now and we can always move that bit out in a follow-up if we decide that it's better.

Yes. Let me rethink that after implementing `file(pattern)` revset. Since revset error can be reported with a span, an unmodified `pattern` string might not be needed. Then, it might be better to move the error construction back to cli layer.

File: lib/src/repo_path.rs (hunk 1 above)
Comment (yuja) [reply lib/src/repo_path.rs]: sure, done.

**Pull request merged**